### PR TITLE
Incoming avoidance

### LINKED ISSUE #22: Incoming_avoidance
The car will be able to avoid incoming conclusion by moving forward or backwards

## Acceptance Criteria:

- [x] the car will have an absolute minor threshold for surrounding objects
- [x] Ensure separation for absolute minor threshold of external objects and a threshold for when the car should stop for stationary objects
- [x] Reverse movement when front of simulator car hits absolute threshold
- [x] Forward movement when back of simulated car hits absolute threshold
- [x] Speed increased when distance between absolute minor threshold and car current location becomes smaller

### PR DESCRIPTION
**Pull request by liamaxelrod**
This will finish off the milestone of incoming avoidance since it is now fulfilled all criteria

So I'm making a pull request to 
close #22 

### :ballot_box_with_check: Definition of Done checklist
- [x] Meaningful title and valuable description in pull request
- [x] Code change tested manually if not covered by automated tests
- [x] Pull request linked to one or more issues/stories
- [x] Acceptance criteria of the linked stories satisfied
- [x] Code change comprehended by those who approved it
- [x] State of target branch maintained or improved after the code change

### CODE REVIEW

File: arduino/smartcar/smartcar.ino [hunk 1]
```diff
@@ -11,8 +11,11 @@
 MQTTClient mqtt;
 WiFiClient net;
 
+bool safetyFeatures = false;
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino]: Add a comment, what this variable does/represent

File: arduino/smartcar/smartcar.ino [hunk 2]
```diff
@@ -133,10 +136,16 @@ void loop()
             const auto rightIRDis = rightIR.getDistance();
             const auto backIRDis = backIR.getDistance();
             
-            if (safetySystem){ //check if the safety system is enabled
+
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino:139]: You removed a useful comment here, maybe replace it with another comment.

File: arduino/smartcar/smartcar.ino [hunk 3]
```diff
@@ -155,7 +164,68 @@ void loop()
...
+            }
+            canDrive = false;//so the car can move in the stop soon
+        } else {
+            canDrive = true;//so the car will stop again if it hits the stop zone
+        }
+}
+
+void incomingAvoidanceThreshold(long frontUltDis, long frontIRDis, long backIRDis)
+{
+    if (frontUltDis <= 30 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0)//Ford threshold 1
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino:186]: Ford threshold 1? Maybe add more details to this comment so it is easier to understand. 
Add more to the other if else conditionals too.

File: arduino/smartcar/smartcar.ino [hunk 4]
```diff
@@ -155,7 +164,68 @@ void loop()
     
 }
 
-//If the connection breaks, this method will be called
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino:158]: You had removed one of my comments and put your code between my comment and my method. Just be careful when you add code that it doesn't affect other methods/functions.

File: arduino/smartcar/smartcar.ino [hunk 5]
```diff
@@ -155,7 +164,68 @@ void loop()
...
+        Serial.println("moving forward level 3");
+        activeAvoidance = true;
+    } else if (frontUltDis == 0 && frontIRDis == 0 && backIRDis == 0 && activeAvoidance)
+    {
+        car.setSpeed(0);
+        activeAvoidance = false;
+    } 
+}
+
+//This method will be called when the connection breaks from the broker
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino:229]: I added my comment back here above my method

File: arduino/smartcar/smartcar.ino [hunk 6]
```diff
@@ -155,7 +164,68 @@ void loop()
...
+void stopZoneAutoBreak(long frontUltDis, long frontIRDis, long backIRDis)
+{
+     if (frontUltDis <= 100 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0 || backIRDis <= 40 && backIRDis != 0)//stop zone
+         {
+            if (canDrive)//check whether you're in the stop zone
+            {
+                car.setSpeed(0);
+                Serial.println("Emergency stop 1");
+            }
+            canDrive = false;//so the car can move in the stop soon
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino:177]: You wrote soon here, I assume you mean Zone ;)

File: arduino/smartcar/smartcar.ino [hunk 7]
```diff
@@ -155,7 +164,68 @@ void loop()
...
+        car.setSpeed(-90);
+        Serial.println("backing up level 1");
+        activeAvoidance = true;
+    } else if (frontUltDis <= 60 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0)//Ford threshold 2
+    {
+        car.setSpeed(0);
+        car.setSpeed(-60);
+        Serial.println("backing up level 2");
+        activeAvoidance = true;
+    } else if (frontUltDis <= 90 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0)//Ford threshold 3this
```
Comment (AlbinKarlsson) [inline arduino/smartcar/smartcar.ino]: At the end of you comment here you wrote 3this. Not sure what you wanted to write here. 
Add more details like i mentioned in the previous comment

**Review by AlbinKarlsson — CHANGES_REQUESTED** [review]
I have added some comments. The code looks good, however I noticed a few things. Please revise! :D

File: arduino/smartcar/smartcar.ino (hunk 1 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino]: Implemented comment

File: arduino/smartcar/smartcar.ino (hunk 2 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino:139]: Added the comment

File: arduino/smartcar/smartcar.ino (hunk 4 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino:158]: I see that but at the end I did put a new comment with your last well what method

File: arduino/smartcar/smartcar.ino (hunk 6 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino:177]: Thank you for noticing that

File: arduino/smartcar/smartcar.ino (hunk 7 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino]: It's fixed now

File: arduino/smartcar/smartcar.ino (hunk 5 above)
Comment (liamaxelrod) [reply arduino/smartcar/smartcar.ino:229]: Thank you I appreciate that

**Comment by liamaxelrod** [comment]
There's been a couple of problems during this pull request so I'm gonna start fresh from a new pull request

**Pull request closed without merging**